002-es6-modules.md

```diff
@@ -193,110 +193,69 @@ The abstractjob operation NodeModuleEvaluationJob with parameters `source` and
-`bar.mjs`. Since `bar.js` is not the listed entry point this was considered
-okay. If foo was providing this file explicitly like `lodash` has this can be
-mitigated easily by using as proxy module should `foo` choose to provide one:
-
-```js
-Object.defineProperty(module, 'exports', {
-  get() {
-    return require(__filename.replace(/\.js$/,'.mjs'))
-  },
-  configurable:false
-});
-Object.freeze(module);
-```
-
-It is recommended going forward that developers not rely on the file extensions
-in packages they do not control.
-
-### 5.1.2. Ecosystem Concerns
-
-Concerns of ecosystem damage when using a new file extension were considered as
-well. Firewall rules and server scripts using `*.js` as the detection mechanism
-for JavaScript would be affected by our change, largely just affecting
-browsers. However, new file extensions and mimes continue to be introduced and
-supported by such things. If a front-end is unable to make such a change, using
-a rewrite rule from `.mjs` to `.js` should sufficiently mitigate this.
-
-There were proposals of using `package.json` as an out of band configuration as
-well.
-
-* removes one-off file execution for quick scripting like using `.save` in the
-  repl then running.
-* causes editors/asset pipelines to have knowledge of this. most work solely on
-  file extension and it would be prohibitive to change.
-    * HTTP/2 PUSH solutions would also need this and would be affected
-* no direction for complete removal of CJS. A file extension leads to a world
-  without `.js` and only `.mjs` files. This would be a permanent field in
-  `package.json`
-* per file mode requirements mean using globs or large lists
-    * discussion of allowing only one mode per package was a non-starter for
-      migration path issues
-* `package.json` is not required in `node_modules/` and raw files can live in
-  `node_modules/`
-    * e.g. `node_modules/foo/index.js`
-    * e.g. `node_modules/foo.js`
-* complex abstraction to handle symlinks used by tools like
-  [link-local](https://github.com/timoxley/linklocal)
-    * e.g. `node_modules/foo -> ../app/components/foo.js`
+Require that Module source text has at least one `import` or `export` declaration.
+A module with only an `import` declaration and no `export` declaration is valid.
+Modules, that do not export anything, should specify an `export {}` to make their
+intentions clear and avoid accidental parse errors while removing `import`
+declarations.
+
+A package opts-in to the Module goal by specifying `"module"` as the parse goal
+field *(name not final)* in its `package.json`. Package dependencies are not
+affected by the opt-in and may be a mix of CJS and ES module packages. If a parse
+goal is not specified, then attempt to parse source text as the preferred goal.
+If there is a parse error that may allow another goal to parse, then parse as
+the other goal, and so on. After this, the goal is known unambiguously and the
```

Comment:
there have been some implications we could encounter more parse goals down the line https://twitter.com/awbjs/status/744701901560651777